### PR TITLE
Added SystemV shared memory based CacheItemPool

### PR DESCRIPTION
**Pull request by tmatsuo**
@dwsupplee Can you also take a look?

### CODE REVIEW

**Comment by tmatsuo** [comment]
See https://github.com/GoogleCloudPlatform/google-cloud-php/issues/989 for context

**Comment by tmatsuo** [comment]
Oops. Sad, but `finally` is not supported with PHP 5.4, and the test is failing

File: src/Cache/SysVCacheItemPool.php [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getItem($key)
+    {
+        $this->loadItems();
+        return current($this->getItems([$key]));
+    }
+    /**
```
Comment (bshaffer) [inline src/Cache/SysVCacheItemPool.php]: nit but, let's consistently put a newline between the ending curly of one function and the start of the next one (several times below as well).

File: tests/Cache/SysVCacheItemPoolTest.php [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+        foreach ($keys as $key) {
+            $item = $this->pool->getItem($key);
+            $item->set('value');
+            $this->pool->saveDeferred($item);
+        }
+
+        $this->assertTrue($this->pool->commit());
+        $this->assertTrue($this->pool->hasItem($keys[0]));
+        $this->assertTrue($this->pool->hasItem($keys[1]));
```
Comment (bshaffer) [inline tests/Cache/SysVCacheItemPoolTest.php]: Test to verify the value is accurate? e.g.

```php
 $this->assertEquals(
    $items[1]->get(),
    $this->pool->getItem($keys[1])->get()
);
```

File: src/Cache/SysVCacheItemPool.php [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+     *           the shared memory. **Defaults to** 1.
+     *     @type string $proj The project identifier for ftok. This needs to
+     *           be a one character string. **Defaults to** 'A'.
+     *     @type int $memsize The memory size in bytes for shm_attach.
+     *           **Defaults to** 10000.
+     *     @type int $perm The permission for shm_attach. **Defaults to** 0600.
+     */
+    public function __construct($options = [])
+    {
+        if (! extension_loaded('sysvshm')) {
```
Comment (bshaffer) [inline src/Cache/SysVCacheItemPool.php]: more standard to not have a space after the `!`

File: src/Cache/SysVCacheItemPool.php [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+     * @var array
+     */
+    private $options;
+
+    /**
+     * Save the current items.
+     *
+     * @return bool true when success, false upon failure
+     */
+    private function saveCurrentItems()
```
Comment (dwsupplee) [inline src/Cache/SysVCacheItemPool.php]: A good practice is to following the following flow of method defintions:

- public
- protected
- private

**Comment by dwsupplee** [comment]
Ack, I had some notes for this and am testing the class - especially against the Spanner session cache.

File: src/Cache/SysVCacheItemPool.php [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getItems(array $keys = [])
+    {
+        $this->loadItems();
+        $items = [];
+        foreach ($keys as $key) {
+            $items[$key] = $this->hasItem($key) ?
```
Comment (dwsupplee) [inline src/Cache/SysVCacheItemPool.php]: A nice standard for ternary statements is

```php
$items[$key] = $this->hasItem($key)
    ? clone $this->items[$key]
    : new Item($key);
```

File: src/Cache/SysVCacheItemPool.php [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+     *
+     *     @type int $variableKey The variable key for getting the data from
+     *           the shared memory. **Defaults to** 1.
+     *     @type string $proj The project identifier for ftok. This needs to
+     *           be a one character string. **Defaults to** 'A'.
+     *     @type int $memsize The memory size in bytes for shm_attach.
+     *           **Defaults to** 10000.
+     *     @type int $perm The permission for shm_attach. **Defaults to** 0600.
+     */
+    public function __construct($options = [])
```
Comment (dwsupplee) [inline src/Cache/SysVCacheItemPool.php]: Why not use an array typehint?

**Comment by bshaffer** [comment]
sorry @dwsupplee !!! I should have reached out to you before merging. Good suggestions. Please let us know if any issues arise from Spanner and we can do a patch release.

**Comment by dwsupplee** [comment]
No worries! Re: Spanner - will do! So far, so good in my testing.

**Pull request merged**